### ISSUE TITLE
iree-translate fails to translate simple_embedding_test.mlir

### ISSUE THREAD

**Issue by marbre** [google/iree#496]
`iree-translate`, build with CMake in Debug mode, finally supports `-iree-mlir-to-vm-bytecode-module`. Thus, I tried to build the simple_embedding sample by enabling

https://github.com/google/iree/blob/016756629a765cc377ede5ce0f4e206278a036a3/iree/samples/CMakeLists.txt#L15

This currently fails with the following:
```
[100%] Generating simple_embedding_test_bytecode_module.module
iree-translate: /home/mbrehler/repos/iree/third_party/llvm-project/mlir/lib/IR/SymbolTable.cpp:221: static mlir::Operation *mlir::SymbolTable::lookupSymbolIn(mlir::Operation *, llvm::StringRef): Assertion `symbolTableOp->hasTrait<OpTrait::SymbolTable>()' failed.
Stack dump:
0.      Program arguments: ../../tools/iree-translate -iree-mlir-to-vm-bytecode-module /home/mbrehler/repos/iree/iree/samples/simple_embedding/simple_embedding_test.mlir -o simple_embedding_test_bytecode_module.module 
../../tools/iree-translate[0x1359509]
../../tools/iree-translate[0x13596b9]
../../tools/iree-translate[0x1357e46]
../../tools/iree-translate[0x1359e4b]
/lib/x86_64-linux-gnu/libpthread.so.0(+0x12890)[0x7f600da88890]
/lib/x86_64-linux-gnu/libc.so.6(gsignal+0xc7)[0x7f600d122e97]
/lib/x86_64-linux-gnu/libc.so.6(abort+0x141)[0x7f600d124801]
/lib/x86_64-linux-gnu/libc.so.6(+0x3039a)[0x7f600d11439a]
/lib/x86_64-linux-gnu/libc.so.6(+0x30412)[0x7f600d114412]
../../tools/iree-translate[0x124ab31]
../../tools/iree-translate[0xd71750]
../../tools/iree-translate[0xd70275]
../../tools/iree-translate[0xd6eb5b]
../../tools/iree-translate[0xd27fdd]
../../tools/iree-translate[0xd29f53]
../../tools/iree-translate[0xd29ee3]
../../tools/iree-translate[0xd2ce99]
../../tools/iree-translate[0xd78305]
../../tools/iree-translate[0xd76416]
../../tools/iree-translate[0xd75806]
../../tools/iree-translate[0x6320ad]
../../tools/iree-translate[0x632dc7]
../../tools/iree-translate[0x634591]
../../tools/iree-translate[0x112aab2]
../../tools/iree-translate[0x112acd9]
../../tools/iree-translate[0x112b21a]
../../tools/iree-translate[0x1469c90]
../../tools/iree-translate[0x146811c]
../../tools/iree-translate[0x1467f0c]
../../tools/iree-translate[0x422e95]
../../tools/iree-translate[0x4218b5]
../../tools/iree-translate[0x42163d]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xe7)[0x7f600d105b97]
../../tools/iree-translate[0x4210fa]
Aborted (core dumped)
```

**Comment by marbre** [comment]
This looks different after applying the changes introduced with PR #518: 

```
/home/mbrehler/repos/iree/iree/samples/simple_embedding/simple_embedding_test.mlir:4:3: error: custom op 'return' is unknown
  return %0 : tensor<4xf32>
  ^
```
It''s similar for the custom modules example, which states that `%c1 = constant 1 : i32` is unknown. I tried to indentify which dep is missing, but couldn't resolve this yet.

**Comment by ScottTodd** [comment]
If it's just a matter of missing a dep or `ALWAYSLINK` somewhere, the [bazel_to_cmake](https://github.com/google/iree/blob/bazel-to-cmake/build_tools/scripts/bazel_to_cmake.py) script should be well suited for helping.

1) Add no-op support for any missing function definitions
2) Run the script across iree/ (or just iree/compiler/ and iree/tools/ ?)
3) Manually revert all changes other than those to DEPS
4) Check if that helped

The fix may be simple here, but hopefully we can lean on the tools more so it isn't such a manual process to keep each combination of [input, tool, options] working across environments.

**Comment by ScottTodd** [comment]
I just saw a similar `error: custom op 'return' is unknown` error bringing the vk_graphics_integration sample back up on CMake (Bazel works without that error).

**Comment by marbre** [comment]
Just updating the deps assisted by bazel_to_cmake wasn't sufficient. Changes introduced with PR #527 do not resolve the issue. However, it at least revealed some additional requirements for bazel_to_cmake...

**Comment by marbre** [comment]
This is not only a problem with `iree-translate`, `iree-opt` fails with similar errors. Since this is not an issue with the Bazel build, adding no-op support shouldn't be the solution. Some further thoughts below. The TL;DR is that it might be an issue with tablegen.

One difference is who tablegen libraries are created. E.g. the configuration for the `FlowEnumsGen` targets looks as follows in Bazel:

```Bazel
gentbl(
    name = "FlowEnumsGen",
    tbl_outs = [
        ("-gen-enum-decls", "FlowEnums.h.inc"),
        ("-gen-enum-defs", "FlowEnums.cpp.inc"),
    ],
    tblgen = "@llvm-project//mlir:mlir-tblgen",
    td_file = "FlowBase.td",
    td_srcs = [
        ":td_files",
        "//iree/compiler/Dialect/IREE/IR:td_files",
        "@llvm-project//mlir:OpBaseTdFiles",
    ],
)
```
The *equivalent* CMake configuration is
```cmake
iree_tablegen_library(
  NAME
    FlowEnumsGen
  SRCS
    FlowBase.td
  OUTS
    -gen-enum-decls FlowEnums.h.inc
    -gen-enum-defs FlowEnums.cpp.inc
)
```
The main difference is that Bazel includes additional sources, here the ones in the filegroups `//iree/compiler/Dialect/IREE/IR:td_files` and `@llvm-project//mlir:OpBaseTdFiles`. Taking a quick look, it seems that the generated `FlowEnums.{h,cpp}.inc` files are identical in Bazel as well as in the CMake build. If I got it right. the additional tdfiles are included in CMake in the appropriate IR target. For the considered case, `iree::compiler::Flow::IR`, which also includes the generated files in `SRCS` and `HDRS`, the dependencies are `MLIRIR` and `iree::compiler::Dialect::IREE::IR`.
Also this seems correct, I can image that something is just not working as expected here...

**Comment by marbre** [comment]
Does not seems to be an issue with the dependencies. Did another iteration updating all deps, no change. So no success so far here :(

@ScottTodd I assume the vulkan example is still broken?

**Comment by ScottTodd** [comment]
Yep. The Vulkan sample still fails to build with `error: custom op 'return' is unknown` when running `iree-translate`.

**Comment by ScottTodd** [comment]
Was `iree-translate` able to run correctly at some point before? If so, we should be able to bisect through the git history with a simple input file like https://github.com/google/iree/blob/master/iree/compiler/Translation/test/smoketest.mlir.

It's weird that standard ops are considered unknown. TableGen could be related...

**Comment by marbre** [comment]
```
a3f33f6d082f7f74a46c92439e152a8f2de899fc is the first bad commit
commit a3f33f6d082f7f74a46c92439e152a8f2de899fc
Author: Ben Vanik <benvanik@google.com>
Date:   Thu Jan 16 14:23:21 2020 -0800

    Making -print-ir-after-all work in iree-translate.
    
    PiperOrigin-RevId: 290144895

:040000 040000 efc4793961e6e888e254a5daa888648cfb3de218 af4ca4f0334b83a47f582450a2e02d1f813c1537 M      iree
```

**Comment by marbre** [comment]
Super stupid mistake introduced with 5975d67...
https://github.com/google/iree/blob/de000ea73b2738ad94ebd247d80d2c84a0d14ef5/build_tools/cmake/iree_alwayslink.cmake#L38
`LIB` must be `${LIB}`. Therefore, the alwayslink property was not set at all. Fixing this, I noticed that we have aliased targets in the list, namely `tensorflow::mlir_xla`. The corresponding non aliased target is `tensorflow_mlir_xla`.

I am actually not satisfied with `iree_alwayslink.cmake` after all. After refactoring, I will send a PR including the fix.
